### ISSUE TITLE
Delete, Upsert and Put

### ISSUE THREAD

**Issue by hraberg** [xtdb/xtdb#22]
We want to be able to transact in a few different ways.

Delete can delete either an entire entity, or an attribute.
Upsert merges a new version of an entity with the existing one.
Put overwrites and entity with a new version.

It should preferably be possible to provide an identity attribute when transacting and have the entity id resolved if it an entity with this attribute exists.

**Comment by hraberg** [comment]
One way to identify an existing entity could be via using lookup refs for the id.

**Comment by hraberg** [comment]
This needs to be encoded inside the transaction itself, as you will want to mix and match.

**Comment by hraberg** [comment]
One approach when using a separate document store is something like this:

```clj
[:crux.tx/put :http://dbpedia.org/resource/Pablo_Picasso
"090622a35d4b579d2fcfebf823821298711d3867"
#inst "2018-05-18T09:20:27.966-00:00"]

[:crux.tx/cas :http://dbpedia.org/resource/Pablo_Picasso
"090622a35d4b579d2fcfebf823821298711d3867"
"048ebba27e1da223ce97dded59d46e069ddf921b"
#inst "2018-05-18T09:21:31.846-00:00"]

[:crux.tx/delete :http://dbpedia.org/resource/Pablo_Picasso 
inst "2018-05-18T09:21:52.151-00:00"]
```

The keys can also be hashed to keep the transaction log anonymous:
```clj
[:crux.tx/put "9049023351ca330419c5c5072948a305343c8d91"
"090622a35d4b579d2fcfebf823821298711d3867"
#inst "2018-05-18T09:20:27.966-00:00"]
```

**Comment by hraberg** [comment]
This is partly implemented in #27 but lacks any ingestion mechanism on the Kafka side.

**Comment by hraberg** [comment]
We now also have ingestion via #39, but there's no way to upsert/merge things at the moment, as the user needs to read and merge themselves via compare-and-set right now. It's not entirely clear how these deltas would actually work in the doc store / tx separation.
